lib/StaticAnalyzer/Checkers/SpecialReturnValueStatisticsCollector.cpp

```diff
@@ -0,0 +1,246 @@
+    ID.AddBoolean(checkedForNegative);
+    ID.AddBoolean(checkedForNull);
+  }
+};
+
+std::map<const CallExpr *, std::pair<bool, bool>> CheckedCalls;
+
+class SpecialReturnValueStatisticsCollector
+    : public Checker<check::PostCall, check::PostStmt<BinaryOperator>,
+                     check::DeadSymbols, check::EndOfTranslationUnit> {
+
+  std::unique_ptr<BugType> StatisticsBugType;
+  AnalysisDeclContext *AC;
+
+  void handleComparison(BinaryOperator::Opcode, SymbolRef Sym, const SVal &Val,
+                        CheckerContext &C) const;
+
+public:
+  SpecialReturnValueStatisticsCollector();
+
+  void checkPostCall(const CallEvent &Call, CheckerContext &C) const;
+  void checkPostStmt(const BinaryOperator *BO, CheckerContext &C) const;
+  void checkDeadSymbols(SymbolReaper &SR, CheckerContext &C) const;
+  void checkEndOfTranslationUnit(const TranslationUnitDecl *TU,
+                                 AnalysisManager &Mgr, BugReporter &BR) const;
+};
+} // namespace
+
+REGISTER_MAP_WITH_PROGRAMSTATE(SpecialReturnValueMap, SymbolRef,
+                               SpecialReturnValue)
+
+SpecialReturnValueStatisticsCollector::SpecialReturnValueStatisticsCollector() {
+  StatisticsBugType.reset(
+      new BugType(this, "Special return value", "Statistics"));
+  StatisticsBugType->setSuppressOnSink(true);
+}
+
+void SpecialReturnValueStatisticsCollector::checkPostCall(
+    const CallEvent &Call, CheckerContext &C) const {
+  const auto *Func = dyn_cast_or_null<FunctionDecl>(Call.getDecl());
+  if (!Func)
+    return;
+
+  if (Func->getReturnType()->isVoidType())
+    return;
+
+  const auto RetSym = Call.getReturnValue().getAsSymbol();
+  if (!RetSym)
+    return;
+
+  const auto *Orig = dyn_cast_or_null<CallExpr>(Call.getOriginExpr());
+  if (!Orig)
+    return;
+
+  auto State = C.getState();
+  State = State->set<SpecialReturnValueMap>(RetSym,
+                                            SpecialReturnValue::getUsage(Orig));
+  C.addTransition(State);
+}
+
+void SpecialReturnValueStatisticsCollector::checkPostStmt(
+    const BinaryOperator *BO, CheckerContext &C) const {
+  if (!BO->isRelationalOp() && !BO->isEqualityOp())
+    return;
+
+  auto State = C.getState();
+  const auto *LCtx = C.getLocationContext();
+
+  const auto LVal = State->getSVal(BO->getLHS(), LCtx),
+             RVal = State->getSVal(BO->getRHS(), LCtx);
+
+  if (const auto LSym = LVal.getAsSymbol()) {
+    handleComparison(BO->getOpcode(), LSym, RVal, C);
+  }
+  if (const auto RSym = RVal.getAsSymbol()) {
+    handleComparison(BinaryOperator::reverseComparisonOp(BO->getOpcode()), RSym,
+                     LVal, C);
+  }
+}
+
+void SpecialReturnValueStatisticsCollector::checkDeadSymbols(
+    SymbolReaper &SR, CheckerContext &C) const {
+  auto State = C.getState();
+
+  auto SymbolMap = State->get<SpecialReturnValueMap>();
+  for (const auto Sym : SymbolMap) {
+    if (!SR.isLive(Sym.first)) {
+      CheckedCalls[Sym.second.getCall()];
```

Comment:
Multiple lookups are intentional. The first lookup just ensures that there is a map entry for the call. If this entry existed earlier, this statement ha no effect, the value does not change. If it did not exist a new is constructed using the default constructor of the value type, which is false for bool.